### ISSUE TITLE
ECKEY PK structure duplicates itself on every use

### ISSUE THREAD

**Issue by hanno-arm** [ARMmbed/mbedtls#2034]
__Context:__ SSL configurations include PK contexts for trusted CA's and the own cert. Moreover, multiple SSL connections may concurrently use a single SSL configuration. Hence, any PK context must implement a measure to guarantee thread-safety. For example, the RSA PK context is guarded by a mutex within the core public and private key routines `mbedtls_rsa_private` and `mbedtls_rsa_public`. This issue is about thread-safety measures available for EC PK contexts.

__Issue 1:__ For the PK type `MBEDTLS_PK_ECKEY`, the verification wrapper in `pk_wrap.c` is defined as follows:
```c
 static int eckey_verify_wrap( void *ctx, mbedtls_md_type_t md_alg,
                       const unsigned char *hash, size_t hash_len,
                       const unsigned char *sig, size_t sig_len )
{
    int ret;
    mbedtls_ecdsa_context ecdsa;

    mbedtls_ecdsa_init( &ecdsa );

    if( ( ret = mbedtls_ecdsa_from_keypair( &ecdsa, ctx ) ) == 0 )
        ret = ecdsa_verify_wrap( &ecdsa, md_alg, hash, hash_len, sig, sig_len );

    mbedtls_ecdsa_free( &ecdsa );

    return( ret );
}
```
The function `mbedtls_ecdsa_from_keypair` essentially copies the ECKEY context `ctx` into the local context `ecdsa` before performing the verify operation. 

There are two questions/issues about this:
 1. Is this meant to be a measure to ensure thread-safety? If so, why was it chosen in favour of a mutex? It seems inefficient.
 1. The function `mbedtls_ecdsa_from_keypair` creates a fresh copy of the group underlying `ctx`, in particular not including precomputed points for the purpose of speeding up comb operations. Hence, each call to the above function needs to re-compute those points.

__Issue 2:__ The PK context of type `MBEDTLS_PK_ECDSA` does not contain any measures for thread-safety. On the other hand, it is actually never used within the library (I removed the corresponding `mbedtls_pk_info_t` structure from the library, and everything compiled and passed the tests, except the PK unit test for this PK structure). As it stands, however, it is theoretically possible for a user to register an `MBEDTLS_PK_ECDSA` PK context for their own private key, hence running into race conditions in a multi-connection setup. 

__Internal Reference:__ IOTSSL-2556.

**Comment by mpg** [comment]
> Is this meant to be a measure to ensure thread-safety? If so, why was it chosen in favour of a mutex? It seems inefficient.

Yes it is. I don't remember why it was done this way instead of using a mutex. By the way, if we did this using a mutex, IMO the mutex should be in the `ecp_group` structure, because that's where the thread-safety issue arises, where on thread by start using `T` when it's already allocated but not yet initialised.

Regarding issue 1 (both sub-points) it's worth taking a step back and noting that the whole handling of precomputed points is poorly designed, which consequences way larger than this case: not only does it create thread-safety issues, it also yield suboptimal speed for larger than necessary memory usage, by having the table live in each copy of `ecp_group`. (For example, think about how each ECDHE exchange in TLS will use a fresh `ecp_group`, resulting in computing and remembering a `T` that's not going to be reused, and same when verifying an ECDSA-based cert chain: each cert's publick key in the chain will have a fresh `ecp_group` with wasted `T`.) See internal ref IOTSSL-1865 for a real improvement (that would also fix the thread-safety issue) and IOTSSL-1864 for a temporary but easy solution.

Regarding issue 2, I agree.

Now considering that (1) the issue ought to be addressed by fixing ECP at some point, and (2) the PK layer in its current from might disappear in favor of PSA crypto at some point, I'm not sure how urgent this would be to fix.
